### PR TITLE
Adds a section with Storage OAuth Tokens to te Profile Page

### PR DESCRIPTION
**Pull request by mereghost**
Implements OP#49434

Adds an extra section over the Access Tokens page, so you can "logout" from a certain storage.

image

### CODE REVIEW

**Comment by mereghost** [comment]
@Kharonus ideally removing a token would trigger a revoke of it on the remote server, but I haven't found a way already implemented to do it. =/

**Comment by Kharonus** [comment]
I'll have a look, in our API we added an endpoint to do so, will check NC API.

**Comment by mereghost** [comment]
> I'll have a look, in our API we added an endpoint to do so, will check NC API.

Cool. I was afraid I was missing something that already existed.

File: app/controllers/my_controller.rb [hunk 1]
```diff
@@ -78,7 +78,27 @@ def change_password
...
+  end
+
+  def delete_storage_token
+    token = OAuthClientToken
+      .preload(:oauth_client)
+      .joins(:oauth_client)
+      .where(user: @user, oauth_client: { integration_type: 'Storages::Storage' }).find_by(id: params[:id])
+
+    if token&.destroy
+      # TODO i18n this
```
Comment (akabiru) [inline app/controllers/my_controller.rb]: Seems like it was solved?
```suggestion
```

Comment (Kharonus) [inline app/controllers/my_controller.rb]: 🔧 remove todo comment

File: app/controllers/my_controller.rb [hunk 2]
```diff
@@ -78,7 +78,27 @@ def change_password
   end
 
   # Administer access tokens
-  def access_token; end
+  def access_token
+    @storage_tokens = OAuthClientToken
+                        .preload(:oauth_client)
+                        .joins(:oauth_client)
+                        .where(user: @user, oauth_client: { integration_type: 'Storages::Storage' })
```
Comment (Kharonus) [inline app/controllers/my_controller.rb:85]: ❓ does this match for `integration_type: Storages::NextcloudStorage`? I'm not sure, what type we enter here. But if someone changes that in future, will it be a problem? Just asking out of curiosity, doesn't think this line needs a change.

File: app/views/my/access_token_partials/_storage_tokens_section.html.erb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+See COPYRIGHT and LICENSE files for more details.
+
+++#%>
+<div class="attributes-group" id="storage-token-section">
+  <div class="attributes-group--header">
+    <div class="attributes-group--header-container">
+      <h3 class="attributes-group--header-text"><%= t("my_account.access_tokens.storages.title") %></h3>
+    </div>
+  </div>
```
Comment (akabiru) [inline app/views/my/access_token_partials/_storage_tokens_section.html.erb:34]: 🟠  Should we use the [`AttributeGroups::AttributeGroupHeaderComponent`](https://github.com/opf/openproject/blob/dev/app/components/attribute_groups/attribute_group_header_component.rb)? AFAIK, we're leaning more towards view components with the primer adoption. Context #12917


```suggestion
  <%= render(AttributeGroups::AttributeGroupHeaderComponent.new(title: t("my_account.access_tokens.storages.title"))) %>
```

File: app/views/my/access_token_partials/_storage_tokens_section.html.erb
Comment (Kharonus) [inline app/views/my/access_token_partials/_storage_tokens_section.html.erb:1]: 📝 Those tables are not yet structured as `table-` and `row-components`? shame, but we don't have to make a lunge right now, as they will get replaced by primer components anyway.

**Review by Kharonus — APPROVED** [review]
Just a couple of comments, and please remove the `TODO` comment, otherwise LGTM.

File: spec/controllers/my_controller_spec.rb [hunk 4]
```diff
@@ -345,5 +345,29 @@
...
+      it 'can remove the token' do
+        expect do
+          delete :delete_storage_token, params: { id: token.id }
+        end.to change(OAuthClientToken, :count).by(-1)
+
+        expect(flash[:info]).to be_present
+        expect(flash[:error]).not_to be_present
+        expect(response).to redirect_to(action: :access_token)
+      end
+    end
```
Comment (akabiru) [inline spec/controllers/my_controller_spec.rb:371]: Nice one 👍🏾 
🟢  Noticed we don't have a feature spec, do you think it's worth repurposing into a feature spec instead? I reckon it simulates a more accurate user journey

**Review by akabiru — COMMENTED** [review]
Looking good to me! I'll leave to @Kharonus for the final approval as you already have a conversation going 🙂 

And yes, I ~stole~ borrowed your [traffic light review system](https://github.com/opf/openproject/pull/13223#pullrequestreview-1547996208) 😄

File: app/controllers/my_controller.rb (hunk 1 above)
Comment (mereghost) [reply app/controllers/my_controller.rb]: 🤦🏼

File: spec/controllers/my_controller_spec.rb (hunk 4 above)
Comment (Kharonus) [reply spec/controllers/my_controller_spec.rb:371]: Everything we can test in unit test I'd rather test in unit tests. Feature specs do have a much bigger ramp up time.

So, I'd say, unless we encounter a nasty regression we couldn't test in a unit test, we should bother with a feature specs in this view.

File: app/controllers/my_controller.rb (hunk 2 above)
Comment (mereghost) [reply app/controllers/my_controller.rb:85]: Good question!

⚠️ I'm citing this out of the top of my mind, so details might be a little off ⚠️ 

`Storages::NextcloudStorage` is implemented as a STI on `Storages::Storage`, meaning that to for the purposes of associations it is a `Storage::Storage`.

If and when we load the association, though, the STI mechanism will instantiate the actual `Storages::SpecializedStorage`.

You can see it in place by running `OAuthClientToken.first.oauth_client.integration`

File: spec/controllers/my_controller_spec.rb (hunk 4 above)
Comment (mereghost) [reply spec/controllers/my_controller_spec.rb:371]: @akabiru I thought about it, but seemed so minor that the overhead of the feature spec would be way to high for such a small piece of code.

@Kharonus agreed.

**Comment by mereghost** [comment]
> And yes, I ~stole~ borrowed your [traffic light review system](https://github.com/opf/openproject/pull/13223#pullrequestreview-1547996208) smile

Oh, I stole it from someone else, feel free to plunder it!

File: app/views/my/access_token_partials/_storage_tokens_section.html.erb (hunk 3 above)
Comment (mereghost) [reply app/views/my/access_token_partials/_storage_tokens_section.html.erb:34]: This page isn't using any components, so I just kept with what was in place. I'm not sure we should update it. =/

**Pull request merged**